FuncSim trap interface, load/store MIPS traps

pukhovv opened the pull request.

pukhovv committed:
Add test MIPS32_instr: beq

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
New unit-tests for ISA (18 implemented) (#530)

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Merge remote-tracking branch 'upsteram/master'

pukhovv committed:
New unit-tests for mips ISA (27 implemented) (#530)

pukhovv committed:
Requested fixes in unit-tests (#530)

pukhovv committed:
Merge remote-tracking branch 'upsteram/master'

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Merge remote-tracking branch 'upsteram/master'

pukhovv committed:
New unit-tests for ISA (40 implemented) (#530)
	Fixes in jmp cases.
	Fixed movn/movz case.
	Test for 64bit instr in 32bit ctor.
	References for bug-issue (#709).

pukhovv committed:
Merge branch 'master' of https://github.com/pooh64/mipt-mips

pukhovv committed:
New unit tests for ISA (55 done) (#530). Splitting

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Some other tests splitted, #709 tests disabled. (#530)

pavelkryukov committed:
Merge branch 'master' into master

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Split tests, fix different misprints (#530)

pukhovv committed:
Merge branch 'master' of https://github.com/pooh64/mipt-mips

pukhovv committed:
Requested fixes in code form (#530)

pukhovv committed:
Change shift-variable overflow tests (#530)

kkorolev committed:
Merge branch 'master' into master

pukhovv committed:
Resolve conflict: unit_test.cpp

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Simple Trap class interface (#602)

pukhovv committed:
fetch upstream/master

pukhovv committed:
Merge submodules

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
(#602) MIPS_TRAP enumenration, new Trap methods

pukhovv committed:
(#602) New trap_types interface prototype

pukhovv committed:
(#602) Last commit fix

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
(#602) Trap class interface

From/to riscv and gdb conversion methods.
Dumping method.

pukhovv committed:
Delete traces module from commit.

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Remove used headers from binutils-gdb

pukhovv committed:
Restore binutils headers and hide internal enum

pukhovv committed:
Implement TrapType to uint8 type conversion

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Add std::hash for TrapType

pukhovv committed:
Add Trap explicit constructors and operators

pukhovv committed:
Fix translate_trap function

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Add addition overflow behaviour, fix old tests

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Restore unit-tests

pukhovv committed:
Fixed addition overflow

pukhovv committed:
Enable addition overflow tests

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Use sign_t in addition overflow

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Move test_addition_overflow to macro.h

pukhovv committed:
Merge branch 'master' of https://github.com/pooh64/mipt-mips

pukhovv committed:
Add subtraction overflow behavior

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Add MIPS ExcCodes and MARS Exceptions

pukhovv committed:
Remove merge dumps from sources

pukhovv committed:
Add new TrapTypes, update submodules

pukhovv committed:
Use Trap::SYSCALL instead of .is_syscall method

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
FuncSim exception handler options

pukhovv committed:
Fix load/store alignment check

pavelkryukov commented on simulator/func_sim/alu.h:
```diff
@@ -41,10 +41,10 @@ struct ALU
     void load_addr_aligned( I* instr) {
         load_addr( instr);
-        if ( instr->mem_addr % 4 != 0)
+        if ( instr->mem_addr % sizeof(T) != 0)
```
It feels that you're reinventing `I::mem_size` here.

pukhovv commented:
SPIM TT and verbose dump showed an error in lhu align check, other instructions produce correct exceptions.
How to implement string knob for trap handling oprions? I think it can be something like that:
"stop_on_halt verbose" but I think it's too complicated.

pavelkryukov commented:
> I think it can be something like that: "stop_on_halt verbose" but I think it's too complicated.

For binary knobs (true/false) you can use our config::Switch structures.
For non-binary knobs (like BP policy) it is preferrable to use strings. If you used numbers, users would be confused and the system would be hardly expandable.

pukhovv commented:
> For non-binary knobs (like BP policy) it is preferrable to use strings. If you used numbers, users would be confused and the system would be hardly expandable.

Oh, I did not think that this interface will be used by users directly.

pukhovv committed:
Fix alu alinged load/store templates

pavelkryukov commented:
> SPIM TT and verbose dump showed an error in lhu align check, other instructions produce correct exceptions.

I don't understand that point, could you please describe more details?

pukhovv commented:
Spim tests generate exception messages. lhu instruction checked the alignment to 32 bits, and it should have been 16 bits. So verbose exception dump showed two redundant exceptions. Now it's fixed.

pavelkryukov commented:
Could you please provide setup via FuncSim constructor so we can have unit tests for the new behaviors?

pavelkryukov commented on simulator/func_sim/func_sim.cpp:
```diff
@@ -104,16 +107,47 @@ Trap FuncSim<ISA>::step_system()
+        if ( trap != Trap::NO_TRAP)
+            return trap;
     }
     return Trap(Trap::NO_TRAP);
```
I feel that we may merge this method with `run_until_trap` now, simply using different simulation configurations.

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Move TrapType-macro commas to trap_types.def

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Add handle-trap FuncSim constuctor options

pukhovv committed:
Merge branch 'master' of https://github.com/pooh64/mipt-mips

pavelkryukov commented:
Please provide tests to cover the new behaviors, you may start with copy-pasting these:
https://github.com/MIPT-ILab/mipt-mips/blob/e7ba8af4399cd9d652c4488c1c43abb7add9c1d7/simulator/func_sim/t/unit_test.cpp#L195-L201

pukhovv committed:
Add configuration for trap handler to simulator

Add trap_mode, _critial, _verbose to config
Update constructors in simulator interface
Remove redundant run_until_trap method

pukhovv committed:
Improve Trap class and handle_trap in FuncSim

Update Trap enums, add new conversion methos.
Add new utests for Trap class and handle_trap_mode

pukhovv committed:
Add new utests for handle_trap

pavelkryukov commented on simulator/func_sim/func_sim.cpp:
```diff
@@ -2,15 +2,29 @@
     , kernel( Kernel::create_dummy_kernel())
+    , handle_trap_mode(handle_trap_str_map.at(trap_mode))
+    , handle_trap_critical(trap_critical)
+    , handle_trap_verbose(trap_verbose)
```
Could you please explain why we have 3 control variables instead of a single variable, at least for ctor arguments?

pukhovv replied:
There are 4 incompatible behaviors described by the string, two options set by switch. Should I make them 6 in one enum?

pavelkryukov replied:
I don't care much about the internals, but it is more convenient to have 1 string argument instead of 1 string argument + 2 Boolean arguments. Boolean arguments are evil since you don't understand their meanings on the caller side (at least in C++)

pavelkryukov commented on simulator/func_sim/func_sim.h:
```diff
@@ -16,6 +16,7 @@
 
 #include <memory>
 #include <string>
+#include <map>
```
Please order includes in alphabetical order.

pavelkryukov replied:
https://github.com/MIPT-ILab/mipt-mips/wiki/Code-style-guidelines#order-of-includes

pavelkryukov commented on simulator/func_sim/t/unit_test.cpp:
```diff
@@ -194,12 +221,34 @@ static auto get_simulator_with_test( const std::string& isa, const std::string&
+TEST_CASE( "Torture_Test: Stop on halt")
+{
+    CHECK( get_simulator_with_test("mips32", valid_elf_file, "stop_on_halt")->run( 1) == Trap::NO_TRAP );
+    auto trap = get_simulator_with_test("mips32", valid_elf_file, "stop_on_halt")->run( 10000);
```
You may put it to a single line
```
CHECK( get_simulator_with_test("mips32", valid_elf_file, "stop_on_halt")->run( 10000) == Trap::HALT)
```

pavelkryukov commented on simulator/func_sim/t/unit_test.cpp:
```diff
@@ -194,12 +221,34 @@ static auto get_simulator_with_test( const std::string& isa, const std::string&
+    CHECK( get_simulator_with_test("mips32", valid_elf_file, "ignore")->run( 1) == Trap::NO_TRAP );
+    auto trap_stop   = get_simulator_with_test("mips32", valid_elf_file, "stop"  )->run( 10);
+    auto trap_ignore = get_simulator_with_test("mips32", valid_elf_file, "ignore")->run( 10);
+    CHECK( trap_stop   != Trap::NO_TRAP );
```
Same here

pavelkryukov commented on simulator/export/gdb/gdb_wrapper.cpp:
```diff
@@ -79,7 +79,7 @@ void GDBSim::resume( uint64 step) try
         trap = cpu->run_single_step ();
     else
-        trap = cpu->run_until_trap( instrs_to_run);
+        trap = cpu->run( instrs_to_run);
```
To maintain behavior, we should change ctor call for GDB wrapped simulator

pavelkryukov commented on simulator/simulator.cpp:
```diff
@@ -18,14 +18,19 @@
     struct Builder {
-        virtual std::unique_ptr<Simulator> get_funcsim( bool log) = 0;
+        virtual std::unique_ptr<Simulator> get_funcsim( bool log,
+            const std::string &trap_mode = "stop_on_halt", bool trap_critical = false, bool trap_verbose = true) = 0;
```
Virtual methods should have no default arguments (see https://google.github.io/styleguide/cppguide.html#Default_Arguments)

pukhovv committed:
Simplify Simulator constructor, update interface

pukhovv committed:
Remove handle_trap_str_map

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pavelkryukov commented:
The handling of the options is extremely over-engineered. I'll merge that but I'll do a massive clean up afterwards.

pukhovv committed:
Remove unnecessary map type from FuncSim

pavelkryukov committed:
Merge branch 'master' into master
